## nn_model_stage2

Quick optimization of initial proof of concept tests, using pandas on a local machine.

In [ ]:
# Import dependencies
import pandas as pd
import numpy as np

from sklearn.cluster import KMeans
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
import tensorflow as tf

import plotly.express as px

In [ ]:
# URL of the CSV file
url = 'https://project4-nyctaxi.s3.us-east-1.amazonaws.com/train.csv'

# Read the CSV file into a DataFrame
df = pd.read_csv(url)

# Display the first few rows of the DataFrame
df.head()


id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N            455  
1         40.731152                  N            663  
2         40.710087                  N           2124  
3         40.706718                  N            429  
4         40.782520                  N            435

In [86]:
df.shape

(1458644, 11)

Create a sample to test model.

In [87]:
# Take only the first 30,000 rows as a sample
df = df[:30000]
df.shape

(30000, 11)

In [88]:
# Check to see if the 'id' repeats
df['id'].nunique()

30000

In [89]:
# Remove id column
df = df.drop(columns= 'id')

The data is not in a format that helps keras, before scaling we'll do some data engineering.

In [90]:
df.dtypes

vendor_id               int64
pickup_datetime        object
dropoff_datetime       object
passenger_count         int64
pickup_longitude      float64
pickup_latitude       float64
dropoff_longitude     float64
dropoff_latitude      float64
store_and_fwd_flag     object
trip_duration           int64
dtype: object

In [91]:
# Convert to datetime format
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])

# Extract features
df['hour'] = df['pickup_datetime'].dt.hour  # Hour of the day (0-23)
df['day_of_week'] = df['pickup_datetime'].dt.dayofweek  # Day of the week (0=Monday, 6=Sunday)
df['month'] = df['pickup_datetime'].dt.month  # Month (1-12)
df['day_of_year'] = df['pickup_datetime'].dt.dayofyear  # Day of the year (1-366)
df['is_weekend'] = df['pickup_datetime'].dt.dayofweek >= 5  # Weekend flag (True/False)

df.head()

vendor_id     pickup_datetime     dropoff_datetime  passenger_count  \
0          2 2016-03-14 17:24:55  2016-03-14 17:32:30                1   
1          1 2016-06-12 00:43:35  2016-06-12 00:54:38                1   
2          2 2016-01-19 11:35:24  2016-01-19 12:10:48                1   
3          2 2016-04-06 19:32:31  2016-04-06 19:39:40                1   
4          2 2016-03-26 13:30:55  2016-03-26 13:38:10                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.982155        40.767937         -73.964630         40.765602   
1        -73.980415        40.738564         -73.999481         40.731152   
2        -73.979027        40.763939         -74.005333         40.710087   
3        -74.010040        40.719971         -74.012268         40.706718   
4        -73.973053        40.793209         -73.972923         40.782520   

  store_and_fwd_flag  trip_duration  hour  day_of_week  month  day_of_year  \
0                  N            455    17            0      3           74   
1                  N            663     0            6      6          164   
2                  N           2124    11            1      1           19   
3                  N            429    19            2      4           97   
4                  N            435    13            5      3           86   

   is_weekend  
0       False  
1        True  
2       False  
3       False  
4        True

We've created more scalable columns: hour, day of the week, month, day of the year, and weekend. Now we need to handle to eact time.

For cyclic features like hour of the day and day of the week, use sine and cosine transformations to retain the cyclical nature.

In [93]:
# Encode hour of the day
df['hour_sin'] = np.sin(2 * np.pi * df['hour'] / 24)
df['hour_cos'] = np.cos(2 * np.pi * df['hour'] / 24)

# Encode day of the week
df['day_of_week_sin'] = np.sin(2 * np.pi * df['day_of_week'] / 7)
df['day_of_week_cos'] = np.cos(2 * np.pi * df['day_of_week'] / 7)

df[['hour', 'hour_sin', 'hour_cos', 'day_of_week', 'day_of_week_sin', 'day_of_week_cos']]


hour  hour_sin      hour_cos  day_of_week  day_of_week_sin  \
0        17 -0.965926 -2.588190e-01            0         0.000000   
1         0  0.000000  1.000000e+00            6        -0.781831   
2        11  0.258819 -9.659258e-01            1         0.781831   
3        19 -0.965926  2.588190e-01            2         0.974928   
4        13 -0.258819 -9.659258e-01            5        -0.974928   
...     ...       ...           ...          ...              ...   
29995    22 -0.500000  8.660254e-01            4        -0.433884   
29996     8  0.866025 -5.000000e-01            3         0.433884   
29997     1  0.258819  9.659258e-01            3         0.433884   
29998    18 -1.000000 -1.836970e-16            5        -0.974928   
29999     3  0.707107  7.071068e-01            4        -0.433884   

       day_of_week_cos  
0             1.000000  
1             0.623490  
2             0.623490  
3            -0.222521  
4            -0.222521  
...                ...  
29995        -0.900969  
29996        -0.900969  
29997        -0.900969  
29998        -0.222521  
29999        -0.900969  

[30000 rows x 6 columns]

Now let's adjust the latitude and longitude columns to make them better for scaling.

In [94]:
# Define a function for haversine distance
def haversine(lat1, lon1, lat2, lon2):
    R = 6371  # Radius of Earth in kilometers
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = np.sin(dlat / 2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2.0)**2
    c = 2 * np.arcsin(np.sqrt(a))
    return R * c

# Apply haversine to calculate distance
df['distance_km'] = haversine(df['pickup_latitude'], df['pickup_longitude'],
                              df['dropoff_latitude'], df['dropoff_longitude'])

df[['pickup_latitude', 'pickup_longitude', 'dropoff_latitude', 'dropoff_longitude', 'distance_km']]


pickup_latitude  pickup_longitude  dropoff_latitude  dropoff_longitude  \
0            40.767937        -73.982155         40.765602         -73.964630   
1            40.738564        -73.980415         40.731152         -73.999481   
2            40.763939        -73.979027         40.710087         -74.005333   
3            40.719971        -74.010040         40.706718         -74.012268   
4            40.793209        -73.973053         40.782520         -73.972923   
...                ...               ...               ...                ...   
29995        40.777569        -73.979042         40.771542         -73.956421   
29996        40.762169        -73.983887         40.753639         -73.974396   
29997        40.755260        -73.994858         40.745785         -73.994324   
29998        40.772476        -73.948959         40.781132         -73.985283   
29999        40.726006        -73.983322         40.729347         -73.993660   

       distance_km  
0         1.498521  
1         1.805507  
2         6.385098  
3         1.485498  
4         1.188588  
...            ...  
29995     2.019304  
29996     1.240406  
29997     1.054611  
29998     3.206423  
29999     0.947062  

[30000 rows x 5 columns]

In [ ]:
# Stack coordinates for clustering
coords = np.vstack((df[['pickup_latitude', 'pickup_longitude']].values,
                    df[['dropoff_latitude', 'dropoff_longitude']].values))

# Apply k-means clustering
kmeans = KMeans(n_clusters=20, random_state=42)
kmeans.fit(coords)

# Assign cluster labels for pickup and dropoff
df['pickup_zone'] = kmeans.predict(df[['pickup_latitude', 'pickup_longitude']])
df['dropoff_zone'] = kmeans.predict(df[['dropoff_latitude', 'dropoff_longitude']])

df[['pickup_latitude', 'pickup_longitude', 'pickup_zone', 'dropoff_zone']]


c:\Users\nbart.DESKTOP-3OF7M8N\anaconda3\envs\dev\lib\site-packages\sklearn\base.py:486: UserWarning: X has feature names, but KMeans was fitted without feature names
  warnings.warn(
c:\Users\nbart.DESKTOP-3OF7M8N\anaconda3\envs\dev\lib\site-packages\sklearn\base.py:486: UserWarning: X has feature names, but KMeans was fitted without feature names
  warnings.warn(


pickup_latitude  pickup_longitude  pickup_zone  dropoff_zone
0            40.767937        -73.982155           19            15
1            40.738564        -73.980415            3            12
2            40.763939        -73.979027           19             0
3            40.719971        -74.010040            0             0
4            40.793209        -73.973053            4             4
...                ...               ...          ...           ...
29995        40.777569        -73.979042            4             1
29996        40.762169        -73.983887           19             8
29997        40.755260        -73.994858            6             6
29998        40.772476        -73.948959            1             4
29999        40.726006        -73.983322            3            12

[30000 rows x 4 columns]

In [96]:
def calculate_bearing(lat1, lon1, lat2, lon2):
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    dlon = lon2 - lon1
    x = np.sin(dlon) * np.cos(lat2)
    y = np.cos(lat1) * np.sin(lat2) - np.sin(lat1) * np.cos(lat2) * np.cos(dlon)
    bearing = np.arctan2(x, y)
    return np.degrees(bearing) % 360

df['bearing'] = calculate_bearing(df['pickup_latitude'], df['pickup_longitude'],
                                   df['dropoff_latitude'], df['dropoff_longitude'])
df[['bearing']]


bearing
0       99.970196
1      242.846232
2      200.319835
3      187.262300
4      179.473585
...           ...
29995  109.376431
29996  139.871262
29997  177.554978
29998  287.479321
29999  293.104008

[30000 rows x 1 columns]

Let's take a look at all the columns to see what to keep before scaling.

In [97]:
df.head()

vendor_id     pickup_datetime     dropoff_datetime  passenger_count  \
0          2 2016-03-14 17:24:55  2016-03-14 17:32:30                1   
1          1 2016-06-12 00:43:35  2016-06-12 00:54:38                1   
2          2 2016-01-19 11:35:24  2016-01-19 12:10:48                1   
3          2 2016-04-06 19:32:31  2016-04-06 19:39:40                1   
4          2 2016-03-26 13:30:55  2016-03-26 13:38:10                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.982155        40.767937         -73.964630         40.765602   
1        -73.980415        40.738564         -73.999481         40.731152   
2        -73.979027        40.763939         -74.005333         40.710087   
3        -74.010040        40.719971         -74.012268         40.706718   
4        -73.973053        40.793209         -73.972923         40.782520   

  store_and_fwd_flag  trip_duration  ...  day_of_year  is_weekend  hour_sin  \
0                  N            455  ...           74       False -0.965926   
1                  N            663  ...          164        True  0.000000   
2                  N           2124  ...           19       False  0.258819   
3                  N            429  ...           97       False -0.965926   
4                  N            435  ...           86        True -0.258819   

   hour_cos  day_of_week_sin  day_of_week_cos  distance_km  pickup_zone  \
0 -0.258819         0.000000         1.000000     1.498521           19   
1  1.000000        -0.781831         0.623490     1.805507            3   
2 -0.965926         0.781831         0.623490     6.385098           19   
3  0.258819         0.974928        -0.222521     1.485498            0   
4 -0.965926        -0.974928        -0.222521     1.188588            4   

   dropoff_zone     bearing  
0            15   99.970196  
1            12  242.846232  
2             0  200.319835  
3             0  187.262300  
4             4  179.473585  

[5 rows x 23 columns]

In [98]:
df.columns

Index(['vendor_id', 'pickup_datetime', 'dropoff_datetime', 'passenger_count',
       'pickup_longitude', 'pickup_latitude', 'dropoff_longitude',
       'dropoff_latitude', 'store_and_fwd_flag', 'trip_duration', 'hour',
       'day_of_week', 'month', 'day_of_year', 'is_weekend', 'hour_sin',
       'hour_cos', 'day_of_week_sin', 'day_of_week_cos', 'distance_km',
       'pickup_zone', 'dropoff_zone', 'bearing'],
      dtype='object')

Drop: 'pickup_datetime', 'dropoff_datetime', 'pickup_longitude', 'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude', 'hour', 'day_of_week'

Get dummies: 'vendor_id', 'store_and_fwd_flag', 'pickup_zone', 'dropoff_zone'

In [99]:
df = df.drop(columns= ['pickup_datetime', 'dropoff_datetime', 'pickup_longitude', 'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude', 'hour', 'day_of_week'])
df.head()

vendor_id  passenger_count store_and_fwd_flag  trip_duration  month  \
0          2                1                  N            455      3   
1          1                1                  N            663      6   
2          2                1                  N           2124      1   
3          2                1                  N            429      4   
4          2                1                  N            435      3   

   day_of_year  is_weekend  hour_sin  hour_cos  day_of_week_sin  \
0           74       False -0.965926 -0.258819         0.000000   
1          164        True  0.000000  1.000000        -0.781831   
2           19       False  0.258819 -0.965926         0.781831   
3           97       False -0.965926  0.258819         0.974928   
4           86        True -0.258819 -0.965926        -0.974928   

   day_of_week_cos  distance_km  pickup_zone  dropoff_zone     bearing  
0         1.000000     1.498521           19            15   99.970196  
1         0.623490     1.805507            3            12  242.846232  
2         0.623490     6.385098           19             0  200.319835  
3        -0.222521     1.485498            0             0  187.262300  
4        -0.222521     1.188588            4             4  179.473585

In [100]:
df.shape

(30000, 15)

In [101]:
df['dropoff_zone'].nunique()

20

In [102]:
df['pickup_zone'].nunique()

19

In [103]:
# List of categorical columns
categorical_cols = ['vendor_id', 'store_and_fwd_flag', 'pickup_zone', 'dropoff_zone']

# Convert integer columns to 'category' dtype
df[categorical_cols] = df[categorical_cols].astype('category')

# Separate the numeric and categorical columns
df_numeric = df.drop(columns=categorical_cols)
df_categorical = df[categorical_cols]

In [104]:
# Get_dummies for categorical data
dummies = pd.get_dummies(df_categorical, drop_first=True)

In [105]:
dummies

vendor_id_2  store_and_fwd_flag_Y  pickup_zone_1  pickup_zone_2  \
0             True                 False          False          False   
1            False                 False          False          False   
2             True                 False          False          False   
3             True                 False          False          False   
4             True                 False          False          False   
...            ...                   ...            ...            ...   
29995         True                 False          False          False   
29996         True                 False          False          False   
29997        False                 False          False          False   
29998        False                 False           True          False   
29999        False                 False          False          False   

       pickup_zone_3  pickup_zone_4  pickup_zone_5  pickup_zone_6  \
0              False          False          False          False   
1               True          False          False          False   
2              False          False          False          False   
3              False          False          False          False   
4              False           True          False          False   
...              ...            ...            ...            ...   
29995          False           True          False          False   
29996          False          False          False          False   
29997          False          False          False           True   
29998          False          False          False          False   
29999           True          False          False          False   

       pickup_zone_7  pickup_zone_8  ...  dropoff_zone_10  dropoff_zone_11  \
0              False          False  ...            False            False   
1              False          False  ...            False            False   
2              False          False  ...            False            False   
3              False          False  ...            False            False   
4              False          False  ...            False            False   
...              ...            ...  ...              ...              ...   
29995          False          False  ...            False            False   
29996          False          False  ...            False            False   
29997          False          False  ...            False            False   
29998          False          False  ...            False            False   
29999          False          False  ...            False            False   

       dropoff_zone_12  dropoff_zone_13  dropoff_zone_14  dropoff_zone_15  \
0                False            False            False             True   
1                 True            False            False            False   
2                False            False            False            False   
3                False            False            False            False   
4                False            False            False            False   
...                ...              ...              ...              ...   
29995            False            False            False            False   
29996            False            False            False            False   
29997            False            False            False            False   
29998            False            False            False            False   
29999             True            False            False            False   

       dropoff_zone_16  dropoff_zone_17  dropoff_zone_18  dropoff_zone_19  
0                False            False            False            False  
1                False            False            False            False  
2                False            False            False            False  
3                False            False            False            False  
4                False            False        

In [106]:
# Convert boolean columns to binary (0/1)
dummies = dummies.astype(int)

In [107]:
# Re-combine the two df
df = pd.concat([df_numeric, dummies], axis=1)


In [108]:
df.columns

Index(['passenger_count', 'trip_duration', 'month', 'day_of_year',
       'is_weekend', 'hour_sin', 'hour_cos', 'day_of_week_sin',
       'day_of_week_cos', 'distance_km', 'bearing', 'vendor_id_2',
       'store_and_fwd_flag_Y', 'pickup_zone_1', 'pickup_zone_2',
       'pickup_zone_3', 'pickup_zone_4', 'pickup_zone_5', 'pickup_zone_6',
       'pickup_zone_7', 'pickup_zone_8', 'pickup_zone_9', 'pickup_zone_10',
       'pickup_zone_11', 'pickup_zone_12', 'pickup_zone_13', 'pickup_zone_15',
       'pickup_zone_16', 'pickup_zone_17', 'pickup_zone_18', 'pickup_zone_19',
       'dropoff_zone_1', 'dropoff_zone_2', 'dropoff_zone_3', 'dropoff_zone_4',
       'dropoff_zone_5', 'dropoff_zone_6', 'dropoff_zone_7', 'dropoff_zone_8',
       'dropoff_zone_9', 'dropoff_zone_10', 'dropoff_zone_11',
       'dropoff_zone_12', 'dropoff_zone_13', 'dropoff_zone_14',
       'dropoff_zone_15', 'dropoff_zone_16', 'dropoff_zone_17',
       'dropoff_zone_18', 'dropoff_zone_19'],
      dtype='object')

In [109]:
df['is_weekend'] = df['is_weekend'].astype(int)

In [110]:
df

passenger_count  trip_duration  month  day_of_year  is_weekend  \
0                    1            455      3           74           0   
1                    1            663      6          164           1   
2                    1           2124      1           19           0   
3                    1            429      4           97           0   
4                    1            435      3           86           1   
...                ...            ...    ...          ...         ...   
29995                1            567      5          148           0   
29996                1            917      5          126           0   
29997                1            232      4          119           0   
29998                1           1065      3           86           1   
29999                1            319      1           22           0   

       hour_sin      hour_cos  day_of_week_sin  day_of_week_cos  distance_km  \
0     -0.965926 -2.588190e-01         0.000000         1.000000     1.498521   
1      0.000000  1.000000e+00        -0.781831         0.623490     1.805507   
2      0.258819 -9.659258e-01         0.781831         0.623490     6.385098   
3     -0.965926  2.588190e-01         0.974928        -0.222521     1.485498   
4     -0.258819 -9.659258e-01        -0.974928        -0.222521     1.188588   
...         ...           ...              ...              ...          ...   
29995 -0.500000  8.660254e-01        -0.433884        -0.900969     2.019304   
29996  0.866025 -5.000000e-01         0.433884        -0.900969     1.240406   
29997  0.258819  9.659258e-01         0.433884        -0.900969     1.054611   
29998 -1.000000 -1.836970e-16        -0.974928        -0.222521     3.206423   
29999  0.707107  7.071068e-01        -0.433884        -0.900969     0.947062   

       ...  dropoff_zone_10  dropoff_zone_11  dropoff_zone_12  \
0      ...                0                0                0   
1      ...                0                0                1   
2      ...                0                0                0   
3      ...                0                0                0   
4      ...                0                0                0   
...    ...              ...              ...              ...   
29995  ...                0                0                0   
29996  ...                0                0                0   
29997  ...                0                0                0   
29998  ...                0                0                0   
29999  ...                0                0                1   

       dropoff_zone_13  dropoff_zone_14  dropoff_zone_15  dropoff_zone_16  \
0                    0                0                1                0   
1                    0                0                0                0   
2                    0                0                0                0   
3                    0                0                0                0   
4                    0                0                0                0   
...                ...              ...              ...              ...   
29995                0                0                0                0   
29996                0                0                0                0   
29997                0                0                0                0   
29998                0                0                0                0   
29999                0                0                0                0   

       dropoff_zone_17  dropoff_zone_18  dropoff_zone_19  
0                    0                0                0  
1                    0                0                0  
2                    0                0                0  
3                    0                0                0  
4                    0                0                0  
...                ...              ...              ...  
29995                0                0           

As further optimization, let's look at where the outliers are in this data and hopefully remove them altogether.

In [ ]:
# Assuming your 'df' DataFrame has a 'trip_duration' column
Q1 = df['trip_duration'].quantile(0.25)
Q3 = df['trip_duration'].quantile(0.75)
IQR = Q3 - Q1

# Defining the outlier range
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter the outliers
outliers = df[(df['trip_duration'] < lower_bound) | (df['trip_duration'] > upper_bound)]

# Visualize the box plot with Plotly Express
fig = px.box(df, y='trip_duration', title='Trip Duration Box Plot')
fig.show()

# Print outliers
outliers.head()

passenger_count  trip_duration  month  day_of_year  is_weekend  hour_sin  \
2                 1           2124      1           19           0  0.258819   
16                1           2316      4          102           0 -0.965926   
24                1           2485      6          155           0  0.866025   
29                1           2331      3           83           0 -0.500000   
33                1           2269      4          108           1 -0.500000   

    hour_cos  day_of_week_sin  day_of_week_cos  distance_km  ...  \
2  -0.965926         0.781831         0.623490     6.385098  ...   
16 -0.258819         0.000000         1.000000     3.428086  ...   
24 -0.500000        -0.433884        -0.900969    20.602575  ...   
29 -0.866025         0.974928        -0.222521     9.939443  ...   
33 -0.866025        -0.781831         0.623490    10.238385  ...   

    dropoff_zone_10  dropoff_zone_11  dropoff_zone_12  dropoff_zone_13  \
2                 0                0                0                0   
16                0                0                1                0   
24                0                0                0                0   
29                0                0                0                0   
33                0                0                0                0   

    dropoff_zone_14  dropoff_zone_15  dropoff_zone_16  dropoff_zone_17  \
2                 0                0                0                0   
16                0                0                0                0   
24                0                0                0                0   
29                0                0                0                0   
33                0                0                0                0   

    dropoff_zone_18  dropoff_zone_19  
2                 0                0  
16                0                0  
24                0                0  
29                0                0  
33                0                0  

[5 rows x 50 columns]

Some of the values here have 10-40 hour taxi ride durations. These are not typical taxi rides. For the sake of the model accuracy we'll simply remove them. The model will hopefully have a more accurate prediction of regular tax trip durations.

In [142]:
# Remove the outliers from the original DataFrame
df= df[~df.index.isin(outliers.index)]

# Optionally, check the cleaned DataFrame
df.head()


passenger_count  trip_duration  month  day_of_year  is_weekend  hour_sin  \
0                1            455      3           74           0 -0.965926   
1                1            663      6          164           1  0.000000   
3                1            429      4           97           0 -0.965926   
4                1            435      3           86           1 -0.258819   
5                6            443      1           30           1 -0.500000   

   hour_cos  day_of_week_sin  day_of_week_cos  distance_km  ...  \
0 -0.258819         0.000000         1.000000     1.498521  ...   
1  1.000000        -0.781831         0.623490     1.805507  ...   
3  0.258819         0.974928        -0.222521     1.485498  ...   
4 -0.965926        -0.974928        -0.222521     1.188588  ...   
5  0.866025        -0.974928        -0.222521     1.098942  ...   

   dropoff_zone_10  dropoff_zone_11  dropoff_zone_12  dropoff_zone_13  \
0                0                0                0                0   
1                0                0                1                0   
3                0                0                0                0   
4                0                0                0                0   
5                0                0                0                0   

   dropoff_zone_14  dropoff_zone_15  dropoff_zone_16  dropoff_zone_17  \
0                0                1                0                0   
1                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   
5                0                0                0                0   

   dropoff_zone_18  dropoff_zone_19  
0                0                0  
1                0                0  
3                0                0  
4                0                0  
5                0                0  

[5 rows x 50 columns]

OK, now we can start split, test, train and scale

In [144]:
# Define X and y correctly
y = df['trip_duration'].values  # y should only contain the target column, trip_duration
X = df.drop(columns=['trip_duration']).values  # X contains all features except for trip_duration


In [145]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# Check the shapes
print(X_train.shape)  # Expected output: (22500, n_features)
print(y_train.shape)  # Expected output: (22500,)


(21351, 49)
(21351,)


In [146]:
# Preprocess numerical data for neural network

# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [147]:
# Check the shapes of X_train and y_train
print(X_train.shape)  # Expected output: (22500, n_features)
print(y_train.shape)  # Expected output: (22500,)

(21351, 49)
(21351,)


This is different from the original `nn_model`: we're scaling the target with MinMaxScaler to hopefully reduce the impact of outliers on our predictions.

In [148]:
from sklearn.preprocessing import MinMaxScaler

# Initialize MinMaxScaler for the target variable
target_scaler = MinMaxScaler()

# Scale the target variable (trip_duration) in training and test data
y_train_scaled = target_scaler.fit_transform(y_train.reshape(-1, 1))  # Reshaping to a 2D array
y_test_scaled = target_scaler.transform(y_test.reshape(-1, 1))  # Apply same scaling to test data

# Check the shapes after scaling
print(y_train_scaled.shape)  # Should be (22500, 1)
print(y_test_scaled.shape)  # Should be (22500, 1)


(21351, 1)
(7117, 1)


In [149]:
# Define the model architecture
nn_model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(input_shape=(X_train_scaled.shape[1],)),  # Shape of your input features
    tf.keras.layers.Dense(128, activation='relu'),  # First hidden layer with 128 units and ReLU activation
    tf.keras.layers.Dense(64, activation='relu'),   # Second hidden layer with 64 units and ReLU activation
    tf.keras.layers.Dense(32, activation='relu'),   # Third hidden layer with 32 units and ReLU activation
    tf.keras.layers.Dense(1)  # Output layer with 1 unit (for predicting trip duration)
])

# Compile the model
nn_model.compile(loss=tf.keras.losses.Huber(),   # MSE for regression
                 optimizer='adam',            # Adam optimizer is good for regression tasks
                 metrics=['mae'])             # Mean Absolute Error as a metric for evaluation


c:\Users\nbart.DESKTOP-3OF7M8N\anaconda3\envs\dev\lib\site-packages\keras\src\layers\core\input_layer.py:26: UserWarning:

Argument `input_shape` is deprecated. Use `shape` instead.



In [150]:
# Train the model using the scaled features and scaled target
nn_model.fit(X_train_scaled, y_train_scaled, epochs=50, batch_size=32)

Epoch 1/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 2s 990us/step - loss: 0.0253 - mae: 0.1549
Epoch 2/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.0088 - mae: 0.0960
Epoch 3/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 919us/step - loss: 0.0076 - mae: 0.0909
Epoch 4/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 942us/step - loss: 0.0084 - mae: 0.0886
Epoch 5/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 926us/step - loss: 0.0069 - mae: 0.0855
Epoch 6/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 969us/step - loss: 0.0071 - mae: 0.0854
Epoch 7/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.0069 - mae: 0.0841
Epoch 8/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.0062 - mae: 0.0817
Epoch 9/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.0062 - mae: 0.0823
Epoch 10/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 961us/step - loss: 0.0061 - mae: 0.0813
Epoch 11/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.0062 - mae: 0.0813
Epoch 12/50
668/668 ━━━━━━━━━━━━━━━━━━━━ 1s 942us/step - loss: 0.0059 - mae: 0.0804
Epoch 13/50

In [151]:
# Make predictions on the test set (scaled)
y_pred_scaled = nn_model.predict(X_test_scaled)

# Rescale the predictions back to the original range of trip_duration
y_pred_rescaled = target_scaler.inverse_transform(y_pred_scaled)


223/223 ━━━━━━━━━━━━━━━━━━━━ 0s 919us/step


In [152]:
# Calculate MAE and MSE on the rescaled predictions
from sklearn.metrics import mean_absolute_error, mean_squared_error

mae = mean_absolute_error(y_test, y_pred_rescaled)
mse = mean_squared_error(y_test, y_pred_rescaled)

print(f"Test MAE (in original scale): {mae}")
print(f"Test MSE (in original scale): {mse}")


Test MAE (in original scale): 197.74928519148597
Test MSE (in original scale): 70124.17022648526


The first optimization attempt was to use MinMaxScaler; this improved accuracy by about 35 seconds. This is about a 6% improvement. Not great, but a little better. Next we'll try using the Huber loss function, which is supposed to be better at handling outliers.

The accuracy of the model with the Huber loss function was 458 seconds, this is more than 100 seconds more accurate than the original not optimized model. We've improved accuracy by about 19% at this point. This model could predict a trip duration with about 7.6 minutes accuracy. This is not great but it's getting closer.

Maybe there's a way to eliminate outliers altogether from the df, to focus more on regular traffic predictions. This will require a careful look at the dataset.

After removing the outliers we have an accuracy within 198 seconds (about 3.3 minutes). Considering traffic congestion in the New York City area, 3 minutes is pretty close. From a personal anecdotal perspective, I find a 3 minute margin of error acceptable for predicting trip durations.

Further optimization could be done, maybe with model parameters, but the steps taken so far at optimization have increased accuracy by 368 seconds (6.13 minutes).